### ISSUE TITLE
Improve performance on upload: Notice if nothing changed

### ISSUE THREAD

**Issue by westnordost** [streetcomplete/StreetComplete#4077]
## Current behavior

Whenever the map data changes (download, upload) or the user creates or deletes a local edit, `MapDataWithEditsSource` notifies its listeners about the changed map data. 

Most notably amongst the listeners is the `OsmQuestController` which will re-scan the changed element(s) for quests and in turn triggers the map to refresh displaying the quests. This is done for example when the user answered a quest and it takes relatively long (a few hundred milliseconds up to a second or so).

After an edit has been uploaded, the copy of the original OSM element is updated and the edit is deleted. This triggers `MapDataWithEditsSource` to rebuild the cache (=the local changes applied to the original OSM map data) and in turn triggers notifying the listeners about changed.

## The Issue

If the element the edit has been applied to and uploaded has not been changed in the meantime by someone else, which consistitutes about 99% of the times,

- the previous original OSM element + applied edit
- and the updated OSM element after upload

are identical. In the case that the element didn't change at all, notifying the listeners that the map data changed is superfluous and could theoretically be avoided. Currently it is not avoided, which means not only are the local changes rebuilt on upload of each singular edit (I think this is not avoidable) but that updated element is scanned for quests plus the map being updated as a consequence of scanning for quests. Again, for each singular edit.

At least what could be done is to reduce the number of elements for which the listener is called to those elements that actually changed. This requires to compare the edited elements before `rebuildLocalChanges` has been called with the edited elements after it has been called.

## Challenges

- since edits and OSM elements are managed independently (`MapDataController`, `ElementEditsController`) and thus updating one and deleting the other during upload theoretically constitutes of two separate operations, it is possible that `MapDataWithEditsSource` is actually notified of a change in data twice for the same uploaded element. Needs to be confirmed is this actually the case

## Remarks

If you intend to look into this, post here first because @Helium314 vaguely registered interest to busy himself with it before.

**Comment by westnordost** [comment]
@Helium314, I think you started some implementation in the frame of another ticket already?

I noticed that the solution for this ticket will not only improve performance on upload, but also on the normal usage for most users, as auto-upload after each change is active. So currently after each quest solution/edit for users that have auto-upload enabled, the quest icons and map data (in case of overlays displayed) are updated twice in fast succession, each with usually the same data. Thus, the solution to this ticket will speed up this normal usage of the app by a factor of 2.

**Comment by Helium314** [comment]
> @Helium314, I think you started some implementation in the frame of another ticket already?

No, I had sone nothing on this.
But I just had a look at https://github.com/streetcomplete/StreetComplete/blob/6e4cd3d1351bae9068fb0d2032509fc021030d60/app/src/main/java/de/westnordost/streetcomplete/data/osm/edits/MapDataWithEditsSource.kt#L61
So before `rebuildLocalChanges`, we need to check whether all keys in `deleted` are in `deletedElements`, and whether all elements and geometries in `updated` are already in `updatedElements`/`updatedGeometries`.
i.e.
```kotlin
val nothingChanged = deletedElements.containsAll(deleted) && updated.all { element ->
    val key = ElementKey(element.type, element.id)
    updatedElements[key] == element && updatedGeometries[key] == updated.getGeometry(element.type, element.id)
}
```
Though elements should probably be allowed to be equal except for the timestamp and maybe version.

If all is as expected, there is no need to call `callOnUpdated`... right?

**Comment by westnordost** [comment]
Yes, exactly, on all points, including the note about `version` and `timestamp`. Hence, the code above won't work because IIRC the `version` of the local element is never updated and the `timestamp` will always differ by at least by a few milliseconds.

So, only everything except that should be compared. And if it is the same, the elements in deletedElements/updatedElements/updatedGeometries should be **removed** from `MapDataWithEditsSource` because the most current element is now from `MapDataController`. Plus of course, most importantly,

> If all is as expected, there is no need to call callOnUpdated... right?

this.

**Comment by Helium314** [comment]
> I am not sure if it is necessary to compare the geometry. If the geometry changed, it means some other node changed too, so that should be caught by the first check.

Right, this is probably not necessary.

> And if it is the same, the elements in deletedElements/updatedElements/updatedGeometries should be **removed** from `MapDataWithEditsSource` because the most current element is now from `MapDataController`.

Actually deletedElements/updatedElements/updatedGeometries are cleared as part of `rebuildLocalChanges`, which is why I mentioned that this check should happen before `rebuildLocalChanges`.
And we need to rebuild anyway, because there may be other local changes on those elements. (at least currently I don't see good a way to avoid rebuilding).

When uploading, the check will only find if a single edit was done on that element.
If there were 2 edits to the element (one was just synced, one remains locally), `nothingChanged` will return false because the updated element is missing the second change.
But I think catching this case is probably much more work...

**Comment by westnordost** [comment]
Hmm, good point. So the current solution would only work when auto-upload is on, where every single change is uploaded immediately.

I see two avenues to explore here:
1. Can the `rebuildLocalChanges` be made so that not the complete local changes must be rebuild but a few elements (what has been passed into that function) too? I.e. the local changes are being "re-based"
2. Can at least update call to the listeners be avoided by detecting that none of the elements passed into the function have changed?

**Comment by Helium314** [comment]
>     2. Can at least update call to the listeners be avoided by detecting that none of the elements passed into the function have changed?

That _might_ be simply:
before rebuilding local changes: `val e = updated.map { get(it.type, it.id) }`
right before `callOnUpdated`: `val nothingChanged = e.all { it == get(it.type, it.id) }`

**Comment by westnordost** [comment]
Also, even if it only works if just one edit has been made on an element, in many cases it will still be worth it because even if you have thousands of local edits, they might just as well be each on different elements and not (most) on top of each other on the same element(s)

**Comment by Helium314** [comment]
first version: https://github.com/streetcomplete/StreetComplete/commit/4cdd85835beda00e4be3386af4c12d067abbe5e1
second version: https://github.com/streetcomplete/StreetComplete/compare/master...Helium314:StreetComplete:avoid_calling_listeners

I didn't even test whether it compiles, and may have overlooked something...

**Comment by Helium314** [comment]
Hmm, actually the second check could be done right after `rebuildLocalChanges`, and `return` immediately if nothing was changed.

**Comment by Helium314** [comment]
I did some test:
Store the `somethingMayHaveChanged` result in a companion object, and if it's false `OsmQuestController.onUpdated` logs whether `deleted.isEmpty()` and `previousQuests` and `quests` contain the same keys / positions.
Every time this was logged, both came out true, i.e. no quests changed.

**Comment by westnordost** [comment]
Sorry, you lost me there. I don't understand how this test was relevant. If no elements changed, of course the quests also do not change(?)

**Comment by Helium314** [comment]
The test was
* whether the checks actually ever return that nothing is changed (nothing overlooked)
* that if they detect that nothing has changed, then really no quests change

**Comment by westnordost** [comment]
I recently uploaded 300+ edits, and the performance gain was very noticeable! The counter was like... drdrdrdrdrdrdrdrttt, very satisfying to see!
